components/datetime/src/pattern/item/ule.rs

```diff
@@ -133,9 +134,127 @@ impl AsULE for PatternItem {
     }
 }
 
+/// `GenericPatternItemULE` is a type optimized for efficent storing and
+/// deserialization of `DateTimeFormat` `GenericPatternItem` elements using
+/// the `ZeroVec` model.
+///
+/// The serialization model packages the pattern item in three bytes.
+///
+/// The first bit is used to disriminate the item variant. If the bit is
+/// set, then the value is the `GenericPatternItem::Placeholder` variant. Otherwise,
+/// the `GenericPatternItem::Literal` is used.
+///
+/// In case the discriminant is set:
+///
+/// 1) The rest of the first byte remains unused.
+/// 2) The second byte is unused.
+/// 3) The third byte encodes the placeholder index.
+///
+/// If the discriminant is not set, the bottom three bits of the first byte,
+/// together with the next two bytes, contain all 21 bits required to encode
+/// any [`Unicode Code Point`].
+///
+/// # Diagram
+///
+/// ```text
+/// ┌───────────────┬───────────────┬───────────────┐
+/// │       u8      │       u8      │       u8      │
+/// ├─┬─┬─┬─┬─┬─┬─┬─┼─┬─┬─┬─┬─┬─┬─┬─┼─┬─┬─┬─┬─┬─┬─┬─┤
+/// ├─┴─┴─┼─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┤
+/// │     │          Unicode Code Point             │ Literal
+/// ├─┬───┴─────────────────────────┬───────────────┤
+/// │X│                             │  Placeholder  │ Placeholder
+/// └─┴─────────────────────────────┴───────────────┘
+///  ▲
+///  │
+///  Variant Discriminant
+/// ```
+///
+/// # Optimization
+///
+/// This model is optimized for efficient packaging of the `GenericPatternItem` elements
+/// and performant deserialization from the `GernericPatternItemULE` to `GenericPatternItem` type.
+///
+/// # Constraints
+///
+/// The model leaves at most 8 `PatternItem` variants, and limits the placeholder
+/// to a single u8.
+///
+/// [`Unicode Code Point`]: http://www.unicode.org/versions/latest/
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct GenericPatternItemULE([u8; 3]);
+
+impl GenericPatternItemULE {
+    /// Given the first byte of the three-byte array that `GenericPatternItemULE` encodes,
+    /// the method determins whether the discriminant in
+    /// the byte indicates that the array encodes the `GenericPatternItem::Field`
+    /// or `GenericPatternItem::Literal` variant of the `GenericPatternItem`.
+    ///
+    /// Returns true when it is a `GenericPatternItem::Field`.
+    #[inline]
+    fn determine_field_from_u8(byte: u8) -> bool {
+        byte & 0b1000_0000 != 0
+    }
+
+    #[inline]
+    fn bytes_in_range(value: (&u8, &u8, &u8)) -> bool {
+        match Self::determine_field_from_u8(*value.0) {
+            // ensure that unused bytes are all zero
+            true => *value.0 == 0b1000_0000 && *value.1 == 0,
+            false => {
+                let u = u32::from_be_bytes([0x00, *value.0, *value.1, *value.2]);
+                char::try_from(u).is_ok()
+            }
+        }
+    }
+}
+
+unsafe impl ULE for GenericPatternItemULE {
```

Comment:
Can you add a comment that satisfies the checklist in https://github.com/unicode-org/icu4x/pull/1121/ ?